### PR TITLE
:sparkles: Introduce new resources to transform and apply

### PR DESCRIPTION
**Pull request by aufi**
Extends crane's transform pipeline so that plugins can generate entirely new resources (not just patch or whiteout existing ones).

New resources are split into a minimal skeleton + JSON patch and written to a dedicated new/ directory within each stage, following the same skeleton+patches pattern as transformed resources.

Depends on https://github.com/migtools/crane-lib/pull/151 - merged.

Related to https://github.com/migtools/crane/issues/728

Added adjustment of E2E test suite timeout from 60 to 90 minutes.

## Summary by CodeRabbit

* **New Features**
  * Plugin-generated resources are now supported during transformations.
  * Generated resources are stored separately and combined with existing resources during configuration builds.
  * Generated resources must include valid kind, name, and API version metadata.
  * Resource changes use minimal skeletons and patches for accurate reconstruction.

* **Bug Fixes**
  * Improved handling of resource replacements, duplicates, and whiteouts.

* **Tests**
  * Added coverage for generated resources, validation, directory layouts, and configuration reconstruction.

### CODE REVIEW

**Comment by aufi** [comment]
@coderabbitai review

**Comment by aufi** [comment]
@coderabbitai review

File: internal/transform/writer.go [hunk 1]
```diff
@@ -69,9 +67,13 @@ func (w *KustomizeWriter) WriteStage(artifacts []cranelib.TransformArtifact, for
 	activeResourcesMap := make(map[string]unstructured.Unstructured)
 	var patches []kustomize.Patch
 
+	isNewResourceMap := make(map[string]bool) // tracks IsNewResource per artifact for dedup resolution
+
 	for _, artifact := range artifacts {
 		resourceID := getResourceID(artifact.Resource)
 
+		isNewResourceMap[resourceID] = artifact.IsNewResource
```
Comment (msajidmansoori12) [inline internal/transform/writer.go]: Hi @aufi , 
I found a small bug in how we track whether a resource is “new” during dedup.

When we already have an active plugin-generated resource (IsNewResource=true, should go under new/), and then we see a whiteout duplicate of the same resource, we correctly keep the first one and ignore the whiteout. But before we decide to skip that whiteout, we overwrite the “is new?” flag with the whiteout’s value (false). After we continue, that wrong flag sticks around.

Result: the resource we kept is still the generated one, but it gets written under input/ instead of new/.

Could we only update that flag when an artifact actually wins the dedup (and leave it alone on the “ignore whiteout duplicate” path)?

It would also be great to add a unit test for this: one active new resource + a whiteout duplicate of the same ID, and assert the file lands in new/ (and that kustomization points at new/...).

**Review by msajidmansoori12 — CHANGES_REQUESTED** [review]

File: internal/transform/orchestrator.go [hunk 2]
```diff
@@ -221,16 +229,50 @@ func (o *Orchestrator) transformResources(stage Stage, stagePlugin cranelib.Plug
...
 		artifacts = append(artifacts, artifact)
+
+		// Process new resources generated by the plugin
+		for i, newResource := range response.NewResources {
+			if newResource.GetKind() == "" {
+				return nil, fmt.Errorf("stage %s: new resource #%d missing kind", stage.DirName, i)
+			}
+			if newResource.GetName() == "" {
+				return nil, fmt.Errorf("stage %s: new resource #%d (%s) missing name", stage.DirName, i, newResource.GetKind())
+			}
```
Comment (msajidmansoori12) [inline internal/transform/orchestrator.go:252]: Hi @aufi , small suggestion here — we already reject new resources that are missing kind or name. Would it make sense to also reject missing apiVersion?

Real plugin-generated resources should always have it (I checked crane-lib’s skeleton expects apiVersion too), and failing early here would be clearer than a confusing kustomize/apply error later. Not a hard blocker, just a consistency thought.

File: .github/workflows/e2e-pr-tester.yaml [hunk 3]
```diff
@@ -143,7 +143,7 @@ jobs:
     permissions:
       contents: read
     runs-on: ubuntu-22.04
-    timeout-minutes: 60
+    timeout-minutes: 90
```
Comment (msajidmansoori12) [inline .github/workflows/e2e-pr-tester.yaml:169]: @aufi since we already split the test runs , this can be omitted from this PR. Just a cleanup nit

**Review by msajidmansoori12 — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e-pr-tester.yaml (hunk 3 above)
Comment (aufi) [reply .github/workflows/e2e-pr-tester.yaml:169]: Updated to upstream main file state.

File: internal/transform/orchestrator.go (hunk 2 above)
Comment (aufi) [reply internal/transform/orchestrator.go:252]: Right, added apiVersion check below.

**Review by msajidmansoori12 — APPROVED** [review]
lgtm

**Review by stillalearner — APPROVED** [review]
LGTM

**Comment by aufi** [comment]
[review-docs]

**Comment by Tamar-Dinavetsky** [comment]
## 📚 Documentation Review

Analyzed PR: https://github.com/migtools/crane/pull/766
Latest commit: `536cee3`

Found **4 file(s)** that may need updates:

### 📋 Select files to update

Uncheck any files you do **not** want updated:

- [x] [development/plugin-development.md](https://github.com/migtools/crane/blob/main/docs//development/plugin-development.md): This documentation covers the requirements and implementation patterns for creating custom Crane migration plugins. I suggest updating the documentation to reflect that plugins now support generating entirely new Kubernetes resources in addition to patching existing ones, which requires changing the plugin output format from a simple JSON array to a structured JSON object containing both `patches` and `newResources` fields.
- [x] [development/architecture.md](https://github.com/migtools/crane/blob/main/docs//development/architecture.md): This documentation file provides an architectural overview of the Crane pipeline, detailing the export, transform, apply, and validation phases. I suggest updating the documentation to reflect the new capability allowing plugins to define entirely new resources, which includes the addition of a `new/` directory structure for managing these resources and updated orchestration logic to handle them via skeleton-and-patch splitting.
- [x] [multistage-pipeline.md](https://github.com/migtools/crane/blob/main/docs//multistage-pipeline.md): This documentation outlines the multi-stage Kustomize pipeline workflow in Crane, and I suggest updating it to include the new `new/` directory structure, which allows plugins to generate entirely new resources during the transformation process instead of only modifying existing ones.
- [x] [transform.md](https://github.com/migtools/crane/blob/main/docs//transform.md): This documentation explains the `transform/` directory structure and workflow for Crane's multi-stage Kustomize pipeline. I suggest updating the documentation to reflect the introduction of a new `new/` directory, which will now store minimal skeletons of resources generated by plugins, alongside corresponding updates to the `kustomization.yaml` references and Git best practices to ensure these generated files are correctly tracked and processed.

---

💡 **Next Steps**:
- **Uncheck** any files above that you don't want updated
- When ready, comment `[​update-docs]` to generate a PR with only the checked files
- You can add instructions in your `[​update-docs]` comment:
  - **Global** (first line): `[​update-docs] keep changes minimal, don't add new sections`
  - **Per-file** (next lines): `config-ref.rst: only update the CLI usage example`

*Powered by code-to-docs AI* ✨

**Comment by aufi** [comment]
[update-docs] keep changes minimal, new resources during transformation is not a typical use case, just mention it as possibility

**Comment by Tamar-Dinavetsky** [comment]
## 📚 Documentation Update

Updated **4 file(s)** based on your review selections:

- ✅ `development/plugin-development.md`
- ✅ `transform.md`
- ✅ `development/architecture.md`
- ✅ `multistage-pipeline.md`

### 📄 Changes

#### `development/plugin-development.md`

<details>
<summary><b>View diff</b></summary>

```diff
--- a/development/plugin-development.md
+++ b/development/plugin-development.md
@@ -40,18 +40,27 @@
 
 ### Output (stdout)
 
-A JSON array of RFC 6902 JSONPatch operations:
+Plugins may return a JSON array of RFC 6902 JSONPatch operations. Additionally, plugins can optionally return entirely new resources to be added to the transformation pipeline.
 
 ```json
-[
-  {"op": "remove", "path": "/metadata/uid"},
-  {"op": "remove", "path": "/metadata/resourceVersion"},
-  {"op": "remove", "path": "/status"},
-  {"op": "add", "path": "/metadata/labels/migrated", "value": "true"}
-]
+{
+  "version": "v1",
+  "isWhiteOut": false,
+  "patches": [
+    {"op": "remove", "path": "/metadata/uid"}
+  ],
+  "newResources": [
+    {
+      "apiVersion": "v1",
+      "kind": "ConfigMap",
+      "metadata": {"name": "generated-config", "namespace": "default"},
+      "data": {"key": "value"}
+    }
+  ]
+}
 ```
 
-Return an empty array `[]` if no transformations are needed.
+Return an empty array for `patches` if no transformations are needed. `newResources` is optional; if provided, each resource must contain a valid `kind`, `name`, and `apiVersion`.
 
 ## Writing a Plugin in Go
 
@@ -64,48 +73,21 @@
     "os"
 
     "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+    cranelib "github.com/konveyor/crane-lib/transform"
 )
-
-type PatchOp struct {
-    Op    string      `json:"op"`
-    Path  string      `json:"path"`
-    Value interface{} `json:"value,omitempty"`
-}
 
 func main() {
     var resource unstructured.Unstructured
     if err := json.NewDecoder(os.Stdin).Decode(&resource); err != nil {
-        fmt.Fprintf(os.Stderr, "failed to decode resource: %v\n", err)
         os.Exit(1)
     }
 
-    var patches []PatchOp
-
-    // Example: add a migration label (ensure parent object exists first)
-    patches = append(patches, PatchOp{
-        Op:    "add",
-        Path:  "/metadata/labels",
-        Value: map[string]interface{}{},
-    })
-    patches = append(patches, PatchOp{
-        Op:    "add",
-        Path:  "/metadata/labels/migrated-by",
-        Value: "my-plugin",
-    })
-
-    // Example: remove a specific annotation
-    annotations := resource.GetAnnotations()
-    if annotations != nil {
-        if _, ok := annotations["source-cluster-only"]; ok {
-            patches = append(patches, PatchOp{
-                Op:   "remove",
-                Path: "/metadata/annotations/source-cluster-only",
-            })
-        }
+    // Example response structure
+    response := cranelib.PluginResponse{
+        Patches: []byte(`[{"op": "add", "path": "/metadata/labels/migrated", "value": "true"}]`),
     }
 
-    if err := json.NewEncoder(os.Stdout).Encode(patches); err != nil {
-        fmt.Fprintf(os.Stderr, "failed to encode patches: %v\n", err)
+    if err := json.NewEncoder(os.Stdout).Encode(response); err != nil {
         os.Exit(1)
     }
 }
@@ -121,14 +103,11 @@
 
 ```bash
 #!/bin/bash
-# Simple plugin that adds a label to all resources
-# Note: ensure /metadata/labels exists before adding child keys
-
+# Simple plugin that returns patches
 cat <<EOF
-[
-  {"op": "add", "path": "/metadata/labels", "value": {}},
-  {"op": "add", "path": "/metadata/labels/environment", "value": "production"}
-]
+{
+  "patches": [{"op": "add", "path": "/metadata/labels/environment", "value": "production"}]
+}
 EOF
 ```
 

```

</details>

#### `transform.md`

<details>
<summary><b>View diff</b></summary>

```diff
--- a/transform.md
+++ b/transform.md
@@ -13,6 +13,8 @@
     │   ├── ConfigMap__v1_default_nginx-config.yaml
     │   ├── Deployment_apps_v1_default_wordpress.yaml
     │   └── Service__v1_default_kubernetes.yaml
+    ├── new/
+    │   └── Build_shipwright.io_v1beta1_default_my-app-build.yaml
     ├── patches/
     │   └── default--apps-v1--Deployment--wordpress.patch.yaml
     ├── output/
@@ -24,9 +26,10 @@
 ### What's in Each Directory?
 
 - **`input/`**: Input resources for this stage (flat files)
+- **`new/`**: New resources generated by a plugin during this stage
 - **`patches/`**: Kustomize patches to apply to resources
 - **`output/`**: Materialized output after applying kustomization
-- **`kustomization.yaml`**: Kustomize configuration file (references `input/`)
+- **`kustomization.yaml`**: Kustomize configuration file (references `input/` and `new/`)
 
 ## Working with Transform Output
 
@@ -100,6 +103,7 @@
 transform/
 ├── 10_KubernetesPlugin/     # Stage 1
 │   ├── input/               # Input resources (flat files)
+│   ├── new/                 # Plugin-generated resources
 │   ├── patches/             # Transformations
 │   ├── output/              # Materialized output (organized by namespace)
 │   └── kustomization.yaml
@@ -257,6 +261,10 @@
 - **Cluster-scoped resources**: `Namespace__v1_clusterscoped_default.yaml`
 
 Each file contains a single Kubernetes resource (not multi-document YAML).
+
+### new/
+
+Contains new resources generated by plugins. These are stored as minimal skeleton files, with additional modifications applied via files in the `patches/` directory.
 
 ### patches/
 
@@ -295,6 +303,7 @@
 resources:
 - resources/ConfigMap__v1_default_nginx-config.yaml
 - resources/Deployment_apps_v1_default_wordpress.yaml
+- resources/new/Build_shipwright.io_v1beta1_default_my-app-build.yaml
 - resources/Service__v1_default_kubernetes.yaml
 
 # Whiteout resources are written to resources/ for complete snapshot
@@ -428,6 +437,7 @@
 
 ✅ **Do commit**:
 - `transform/*/input/*.yaml`
+- `transform/*/new/*.yaml`
 - `transform/*/patches/*.yaml`
 - `transform/*/kustomization.yaml`
 

```

</details>

#### `development/architecture.md`

<details>
<summary><b>View diff</b></summary>

```diff
--- a/development/architecture.md
+++ b/development/architecture.md
@@ -37,7 +37,7 @@
 
 1. **Stage discovery** — Scans `transform/` for existing stage directories matching `<number>_<name>` pattern
 2. **Plugin execution** — For plugin-based stages (name ending in `Plugin`), loads and runs the matching plugin to generate JSONPatch operations
-3. **Resource writing** — Writes individual resource files to `<stage>/input/`
+3. **Resource writing** — Writes individual resource files to `<stage>/input/` or `<stage>/new/` (for plugin-generated resources)
 4. **Patch writing** — Writes plugin-generated patches to `<stage>/patches/`
 5. **Kustomization generation** — Generates `kustomization.yaml` linking resources and patches
 
@@ -48,7 +48,7 @@
 
 ### Sequential Consistency
 
-In multi-stage pipelines, each stage runs on the fully materialized output of the previous stage (not raw patches). Each stage contains `input/`, `patches/`, and `output/` directly within the stage directory (e.g., `transform/<stage>/input/`, `transform/<stage>/output/`).
+In multi-stage pipelines, each stage runs on the fully materialized output of the previous stage (not raw patches). Each stage contains `input/`, `new/`, `patches/`, and `output/` directly within the stage directory (e.g., `transform/<stage>/input/`, `transform/<stage>/new/`).
 
 ### Key Components
 
@@ -87,6 +87,8 @@
 3. Are discovered from `~/.local/share/crane/plugins/` (default)
 
 The built-in `KubernetesPlugin` (from `crane-lib`) removes server-managed fields like `metadata.uid`, `metadata.resourceVersion`, `metadata.creationTimestamp`, `metadata.managedFields`, and `status`.
+
+Plugins may optionally generate entirely new resources. These are stored in the `new/` directory within a stage, represented as a minimal skeleton resource and a JSON patch that defines the full object state.
 
 ## PVC Transfer
 
@@ -135,5 +137,5 @@
                     ┌──────────────┐
                     │ Target       │
                     │ Cluster      │
-                    └──────────────┘
+                    └──────┘
 ```

```

</details>

#### `multistage-pipeline.md`

<details>
<summary><b>View diff</b></summary>

```diff
--- a/multistage-pipeline.md
+++ b/multistage-pipeline.md
@@ -19,6 +19,7 @@
 │   │   ├── deployment.yaml          # Grouped by resource type
 │   │   ├── service.yaml
 │   │   └── configmap.yaml
+│   ├── new/                          # Plugin-generated new resources
 │   ├── patches/
 │   │   ├── deployment-myapp-default.yaml
 │   │   └── service-myapp-default.yaml
@@ -36,11 +37,12 @@
 Resources are grouped by type (kind + API group) into multi-document YAML files:
 - Core resources: `deployment.yaml`, `service.yaml`, `pod.yaml`
 - Non-core resources: `route.route.openshift.io.yaml`, `imagestream.image.openshift.io.yaml`
+- New resources: Generated resources are written to the `new/` directory as minimal skeletons and reconstructed using patches.
 
 ### Kustomization File
 
 Each stage contains a `kustomization.yaml` that references:
-- **resources**: List of resource files from the `input/` directory
+- **resources**: List of resource files from the `input/` and `new/` directories
 - **patches**: Strategic merge patches or JSON patches with target selectors
 
 Example:
@@ -49,15 +51,14 @@
 kind: Kustomization
 resources:
 - input/deployment.yaml
-- input/service.yaml
+- new/Build.yaml
 patches:
-- path: patches/deployment-myapp-default.yaml
+- path: patches/Build-my-app-build.yaml
   target:
-    group: apps
-    version: v1
-    kind: Deployment
-    name: myapp
-    namespace: default
+    group: shipwright.io
+    version: v1beta1
+    kind: Build
+    name: my-app-build
 ```
 
 ### Resource Cleanup via Plugins
@@ -75,29 +76,10 @@
 
 **How it works**:
 
-1. **Transform phase**: Plugins analyze exported resources and generate JSONPatch operations
-2. **Patches written**: Operations saved as Kustomize patches in `patches/` directory
-3. **Apply phase**: Embedded kustomize applies patches to resources
-4. **Result**: Clean, declarative manifests ready for target cluster
-
-**Example patch** (auto-generated by kubernetes plugin):
-```yaml
-- op: remove
-  path: /metadata/resourceVersion
-- op: remove
-  path: /metadata/creationTimestamp
-- op: remove
-  path: /metadata/managedFields
-- op: remove
-  path: /status
-- op: remove
-  path: /metadata/uid
-```
-
-This approach ensures:
-- No conflicts with server-managed fields
-- Resources can be applied to any cluster
-- Idempotent operations (safe to re-apply)
+1. **Transform phase**: Plugins analyze exported resources and generate JSONPatch operations. Plugins may also generate entirely new resources.
+2. **Patches written**: Operations and new resource skeletons are saved as Kustomize patches in the `patches/` and `new/` directories.
+3. **Apply phase**: Embedded kustomize applies patches to resources.
+4. **Result**: Clean, declarative manifests ready for target cluster.
 
 ## CLI Usage
 

```

</details>

A docs PR has been created: https://github.com/migtools/crane/pull/806
